bootstrap-app/src/cmr/bootstrap/services/dispatch/core_async_dispatch.clj

```diff
@@ -0,0 +1,143 @@
+    (info "Adding date-time" date-time "to data index channel.")
+    (async/go (>! channel {:date-time date-time}))))
+
+(defn index-collection
+  "Bulk index all the granules in a collection"
+  [this context provider-id collection-id options]
+  (let [channel (:collection-index-channel this)]
+    (info "Adding collection" collection-id "to collection index channel")
+    (async/go (>! channel (merge options
+                                 {:provider-id provider-id
+                                  :collection-id collection-id})))))
+
+(defn index-system-concepts
+  "Bulk index all the tags, acls, and access-groups."
+  [this context start-index]
+  (let [channel (:system-concept-channel this)]
+    (info "Adding bulk index request to system concepts channel.")
+    (async/go (>! channel {:start-index start-index}))))
+
+(defn index-concepts-by-id
+  "Bulk index the concepts given by the concept-ids"
+  [this context provider-id concept-type concept-ids]
+  (let [channel (:concept-id-channel this)]
+    (info "Adding bulk index request to concept-id channel.")
+    (async/go (>! channel {:provider-id provider-id
+                           :concept-type concept-type
+                           :request :index
+                           :concept-ids concept-ids}))))
+
+(defn delete-concepts-from-index-by-id
+  "Bulk delete the concepts given by the concept-ids from the indexes"
+  [this context provider-id concept-type concept-ids]
+  (let [channel (:concept-id-channel this)]
+    (info "Adding bulk delete reqeust to concept-id channel.")
+    (async/go (>! channel {:provider-id provider-id
+                           :concept-type concept-type
+                           :request :delete
+                           :concept-ids concept-ids}))))
+
+(defn bootstrap-virtual-products
+  "Initializes virtual products."
+  [this context provider-id entry-title]
+  (let [channel (:virtual-product-channel this)]
+    (info "Adding message to virtual products channel.")
+    (async/go (>! channel {:provider-id provider-id
+                           :entry-title entry-title}))))
+
+(defrecord CoreAsyncDispatcher
+  [;; Channel for requesting full provider migration
+   provider-db-channel
+   ;; Channel for requesting single collection/granules migration.
+   ;; Takes maps, e.g., {:collection-id collection-id :provider-id provider-id}
+   collection-db-channel
+   ;; Channel for requesting full provider indexing - collections/granules
+   provider-index-channel
+   ;; Channel for processing collections to index.
+   collection-index-channel
+   ;; Channel for processing data newer than a given date-time.
+   data-index-channel
+   ;; Channel for processing bulk index requests for system concepts (tags, acls, access-groups)
+   system-concept-channel
+   ;; channel for processing bulk index requests by concept-id
+   concept-id-channel
+   ;; channel for bootstrapping virtual products
+   virtual-product-channel])
+
+(def dispatch-behavior
```

Comment:
Another really nice thing about this approach is that it's easy to emulate multiple inheritance (only if really needed!)  by simply calling `extend` again with the new parent class, and it's easy to override behaviours in subclasses with a simple `merge` or `assoc`. Extremely powerful -- while maintaining clarity and simplicity -- for multiple implementation scenarios. Which sort of makes it a coding holy grail ... ❤️ Rich Hickey.